src/function/execute.rs

```diff
@@ -37,53 +41,132 @@ where
             })
         });
 
-        // If we already executed this query once, then use the tracked-struct ids from the
-        // previous execution as the starting point for the new one.
-        if let Some(old_memo) = &opt_old_memo {
-            active_query.seed_tracked_struct_ids(&old_memo.revisions.tracked_struct_ids);
-        }
+        let mut iteration_count: u32 = 0;
+        let mut fell_back = false;
 
-        // Query was not previously executed, or value is potentially
-        // stale, or value is absent. Let's execute!
-        let database_key_index = active_query.database_key_index;
-        let id = database_key_index.key_index;
-        let value = match Cycle::catch(|| C::execute(db, C::id_to_input(db, id))) {
-            Ok(v) => v,
-            Err(cycle) => {
+        // Our provisional value from the previous iteration, when doing fixpoint iteration.
+        // Initially it's set to None, because the initial provisional value is created lazily,
+        // only when a cycle is actually encountered.
+        let mut opt_last_provisional: Option<&Memo<<C as Configuration>::Output<'db>>> = None;
+
+        loop {
+            // If we already executed this query once, then use the tracked-struct ids from the
+            // previous execution as the starting point for the new one.
+            if let Some(old_memo) = &opt_old_memo {
+                active_query.seed_tracked_struct_ids(&old_memo.revisions.tracked_struct_ids);
+            }
+
+            // Query was not previously executed, or value is potentially
+            // stale, or value is absent. Let's execute!
+            let mut new_value = C::execute(db, C::id_to_input(db, id));
+            let mut revisions = active_query.pop();
+
+            // Did the new result we got depend on our own provisional value, in a cycle?
+            if C::CYCLE_STRATEGY == CycleRecoveryStrategy::Fixpoint
+                && revisions.cycle_heads.contains(&database_key_index)
+            {
+                let opt_owned_last_provisional;
+                let last_provisional_value = if let Some(last_provisional) = opt_last_provisional {
+                    // We have a last provisional value from our previous time around the loop.
+                    last_provisional
+                        .value
+                        .as_ref()
+                        .expect("provisional value should not be evicted by LRU")
+                } else {
+                    // This is our first time around the loop; a provisional value must have been
+                    // inserted into the memo table when the cycle was hit, so let's pull our
+                    // initial provisional value from there.
+                    opt_owned_last_provisional = self.get_memo_from_table_for(zalsa, id);
```

Comment:
Given we expect this to be a provisional value, it might be good to add a 
```rs
                    debug_assert!(opt_owned_last_provisional
                        .as_ref()
                        .unwrap()
                        .may_be_provisional());
```
for good measure?